src/message_dao.py

```diff
@@ -12,27 +12,141 @@
+        self.connection = connector.connect(
+            os.environ.get("HOST"),
+            "pymysql",
+            user = os.environ.get("USER"),
+            password = os.environ.get("PASSWORD")
+        )
+    
+    def _initialize_tables(self) -> None:
+        db_name = os.environ.get("DB")
+        if not db_name:
+            raise Exception("Missing environment variable for name of database.")
+        with self.connection.cursor() as cursor:
+            cursor.execute("CREATE DATABASE IF NOT EXISTS " + db_name)
+            self.connection.select_db(db_name)
+            # Mapping from message id to message
+            cursor.execute(
+                "CREATE TABLE IF NOT EXISTS Messages"
+                "(message_id VARCHAR(255) NOT NULL,"
+                "reduced_subject VARCHAR(255) NOT NULL,"
+                "from_ VARCHAR(255) NOT NULL,"
+                "in_reply_to VARCHAR(255),"
+                "archive_hash VARCHAR(255) NOT NULL,"
+                "change_id VARCHAR(255),"
+                "lore_link VARCHAR(255),"
+                "PRIMARY KEY (message_id))"
+            )
+            # Mapping from name of state attribute to state
+            cursor.execute(
+                "CREATE TABLE IF NOT EXISTS States"
+                "(state_name VARCHAR(255) NOT NULL,"
+                "value VARCHAR(255) NOT NULL,"
+                "PRIMARY KEY (state_name))"
+            )
+        self.connection.commit()
+
+    def store(self, message: Message) -> None:
+        # Use only the portion of the subject that comes after the brackets that indicate patch information
+        reduced_subject = message.subject.partition("] ")[2].lower()
+        link = lore_link(message.id)
+        sql = "REPLACE INTO Messages VALUES (%s, %s, %s, %s, %s, %s, %s)"
+        with self.connection.cursor() as cursor:
+            cursor.execute(sql, (message.id, reduced_subject, message.from_,
+            message.in_reply_to, message.archive_hash, message.change_id, link))
+        if message.in_reply_to:
+            # Clear cache because the parent's cache is no longer valid: list of children changed
+            self.get.cache_clear()
+        self.connection.commit()
+
+    def _get_children(self, message_id: str) -> List[Optional[Message]]:
+        sql = "SELECT * FROM Messages WHERE in_reply_to=%s"
+        with self.connection.cursor() as cursor:
+            cursor.execute(sql, (message_id,))
+            res = cursor.fetchall()
+        return [self.get(tup[0]) for tup in res]
+
+    @lru_cache
+    def get(self, message_id: str) -> Optional[Message]:
+        sql = "SELECT archive_hash, change_id FROM Messages WHERE message_id=%s"
+        with self.connection.cursor() as cursor:
+            cursor.execute(sql, (message_id,))
+            res = cursor.fetchone()
+        if res is None:
+            return None
+        # Recreate the message object using the archive hash
+        archive_hash = res[0]
```

Comment:
nit: I'd probably write this like

```python
archive_hash, change_id = res[0], res[1]
```

That way us grabbing out the change_id closer to the SELECT statement rather than 10 lines below (line 98 and line 108).

Reply:
Good call! Looks much cleaner that way.